### PR TITLE
Add gradient data to xml exporters

### PR DESCRIPTION
**Pull request by kursatyurt**
## Main changes of this PR

Follow-up of #1315.
Add gradient output support to XML based file formats `vtu` `vtp`. 
Add tests for VTU/VTP export with gradient data.

### CODE REVIEW

File: docs/changelog/1340.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+
+- Added support to export gradient data in XML based exporting, VTU,VTP
+- Scalar gradients are written to vector <ScalarDataname>_gradient `v0_dx,v0_dy,v0_dz,v1_dx ...`
+- Vector gradients are written to vectors <VectorDataname>_dx/dy/dz `v0x_dx,v0y_dx,v0z_dz, ...`
```
Comment (fsimonis) [inline docs/changelog/1340.md:4]: The remaining info should go in the docs

```suggestion
- Added support to export gradient data in XML based exporting, VTU, VTP.
```

**Comment by fsimonis** [comment]
When I run the tests `TestOutput/io/io-VTKExport-ExportDatawithGradient.vtp` cannot be read from paraview as the gradient of the scalar is missing one vertex:

```xml
            <DataArray type="Float64" Name="dataScalar" NumberOfComponents="1" format="ascii">
               1 2 
            </DataArray>
            <DataArray type="Float64" Name="dataScalar_gradient" NumberOfComponents="3" format="ascii">
               1 1 0.0
            </DataArray>
```

**Comment by kursatyurt** [comment]
> When I run the tests `TestOutput/io/io-VTKExport-ExportDatawithGradient.vtp` cannot be read from paraview as the gradient of the scalar is missing one vertex:
> 
> ```
>             <DataArray type="Float64" Name="dataScalar" NumberOfComponents="1" format="ascii">
>                1 2 
>             </DataArray>
>             <DataArray type="Float64" Name="dataScalar_gradient" NumberOfComponents="3" format="ascii">
>                1 1 0.0
>             </DataArray>
> ```

Thank you, there was a logic error it is now fixed.

File: src/io/ExportXML.cpp [hunk 2]
```diff
@@ -196,6 +244,11 @@ void ExportXML::exportData(
     }
     outFile << '\n'
             << "            </DataArray>\n";
+    if (hasGradient && dataDimensions > 1) {
+      exportGradient(data, spaceDim, outFile);
+    } else if (hasGradient && dataDimensions == 1) {
+      exportGradient(data, 1, outFile);
```
Comment (DavidSCN) [inline src/io/ExportXML.cpp:250]: Is the `spaceDim` argument really the spatial dimension? We usually don't support `dim=1` and you also named the variable in the function `spaceDim`. I guess it is rather the data dimension?! If that's the case, couldn't we omit the branch here and just write
```c++
    if (hasGradient) {
      exportGradient(data, dataDimensions, outFile);
    }
```
?

File: src/io/ExportXML.hpp [hunk 3]
```diff
@@ -6,7 +6,7 @@
 #include <vector>
 #include "io/Export.hpp"
 #include "logging/Logger.hpp"
-
+#include "mesh/SharedPointer.hpp"
```
Comment (DavidSCN) [inline src/io/ExportXML.hpp:9]: Newline afterwards.

File: src/io/ExportXML.hpp [hunk 4]
```diff
@@ -94,6 +94,8 @@ class ExportXML : public Export {
   void exportData(
       std::ostream &    outFile,
       const mesh::Mesh &mesh) const;
+
+  void exportGradient(const mesh::PtrData data, const int spaceDim, std::ostream &outFile) const;
```
Comment (DavidSCN) [inline src/io/ExportXML.hpp:98]:
```suggestion
  void exportGradient(const mesh::PtrData data, const int dataDim, std::ostream &outFile) const;
```
?

File: src/io/ExportXML.cpp (hunk 2 above)
Comment (kursatyurt) [reply src/io/ExportXML.cpp:250]: Yes it is dataDimension

File: src/io/ExportXML.hpp (hunk 3 above)
Comment (kursatyurt) [reply src/io/ExportXML.hpp:9]: Fixed as well

**Comment by kursatyurt** [comment]
@DavidSCN both 2D, and 3D tests are ready.

File: src/io/ExportXML.cpp [hunk 5]
```diff
@@ -166,11 +212,13 @@ void ExportXML::exportData(
   }
   outFile << "\n            </DataArray>\n";
 
+  const int spaceDim = mesh.getDimensions();
```
Comment (DavidSCN) [inline src/io/ExportXML.cpp:215]:
```suggestion
```
This variable is unused.

File: src/io/tests/ExportVTPTest.cpp [hunk 6]
```diff
@@ -25,6 +25,66 @@ using namespace precice;
...
 BOOST_AUTO_TEST_SUITE(VTPExport)
 
+BOOST_AUTO_TEST_CASE(ExportDataWithGradient2D)
+{
+  PRECICE_TEST(1_rank)
+  const int dimensions = 2;
+  // Create mesh to map from
+  mesh::Mesh    mesh("MyMesh", dimensions, testing::nextMeshID());
+  mesh::PtrData dataScalar = mesh.createData("dataScalar", 1, 0_dataID, true);
+  mesh::PtrData dataVector = mesh.createData("dataVector", 2, 1_dataID, true);
```
Comment (DavidSCN) [inline src/io/tests/ExportVTPTest.cpp:35]:
```suggestion
  mesh::PtrData dataVector = mesh.createData("dataVector", dimensions, 1_dataID, true);
```

File: src/io/tests/ExportVTPTest.cpp [hunk 7]
```diff
@@ -25,6 +25,66 @@ using namespace precice;
...
+}
+
+BOOST_AUTO_TEST_CASE(ExportDataWithGradient3D)
+{
+  PRECICE_TEST(1_rank)
+  const int dimensions = 3;
+  // Create mesh to map from
+  mesh::Mesh    mesh("MyMesh", dimensions, testing::nextMeshID());
+  mesh::PtrData dataScalar = mesh.createData("dataScalar", 1, 0_dataID, true);
+  mesh::PtrData dataVector = mesh.createData("dataVector", 2, 1_dataID, true);
```
Comment (DavidSCN) [inline src/io/tests/ExportVTPTest.cpp:65]:
```suggestion
  mesh::PtrData dataVector = mesh.createData("dataVector", dimensions, 1_dataID, true);
```

File: src/io/tests/ExportVTUTest.cpp [hunk 8]
```diff
@@ -25,6 +25,68 @@ using namespace precice;
...
 
+BOOST_AUTO_TEST_CASE(ExportDataWithGradient2D)
+{
+  PRECICE_TEST(1_rank)
+
+  // Create mesh to map from
+  int           dimensions = 2;
+  mesh::Mesh    mesh("MyMesh", dimensions, testing::nextMeshID());
+  mesh::PtrData dataScalar = mesh.createData("dataScalar", 1, 0_dataID, true);
+  mesh::PtrData dataVector = mesh.createData("dataVector", 2, 1_dataID, true);
```
Comment (DavidSCN) [inline src/io/tests/ExportVTUTest.cpp:36]:
```suggestion
  mesh::PtrData dataVector = mesh.createData("dataVector", dimensions, 1_dataID, true);
```

File: src/io/tests/ExportVTUTest.cpp [hunk 9]
```diff
@@ -25,6 +25,68 @@ using namespace precice;
...
+}
+
+BOOST_AUTO_TEST_CASE(ExportDataWithGradient3D)
+{
+  PRECICE_TEST(1_rank)
+  int dimensions = 3;
+  // Create mesh to map from
+  mesh::Mesh    mesh("MyMesh", dimensions, testing::nextMeshID());
+  mesh::PtrData dataScalar = mesh.createData("dataScalar", 1, 0_dataID, true);
+  mesh::PtrData dataVector = mesh.createData("dataVector", 2, 1_dataID, true);
```
Comment (DavidSCN) [inline src/io/tests/ExportVTUTest.cpp:66]:
```suggestion
  mesh::PtrData dataVector = mesh.createData("dataVector", dimensions, 1_dataID, true);
```

**Review by DavidSCN — COMMENTED** [review]
I opened the generated test files and the file `io-VTPExport-ExportDatawithGradient3.vtp` does not contain 3D data. See below.

**Comment by kursatyurt** [comment]
> I opened the generated test files and the file `io-VTPExport-ExportDatawithGradient3.vtp` does not contain 3D data. See below.

I see thank you very much.

**Comment by DavidSCN** [comment]
Thanks!

**Pull request merged**